Zotero/Controllers/OpenItemsController.swift

```diff
@@ -0,0 +1,481 @@
+                case pdfKind
+                case noteKind
+                case libraryId
+                case key
+            }
+
+            func encode(to encoder: Encoder) throws {
+                var container = encoder.container(keyedBy: CodingKeys.self)
+                switch self {
+                case .pdf:
+                    try container.encode(true, forKey: .pdfKind)
+
+                case .note:
+                    try container.encode(true, forKey: .noteKind)
+                }
+
+                try container.encode(libraryId, forKey: .libraryId)
+                try container.encode(key, forKey: .key)
+            }
+
+            init(from decoder: Decoder) throws {
+                let container = try decoder.container(keyedBy: CodingKeys.self)
+                let libraryId = try container.decode(LibraryIdentifier.self, forKey: .libraryId)
+                let key = try container.decode(String.self, forKey: .key)
+                if (try? container.decode(Bool.self, forKey: .pdfKind)) == true {
+                    self = .pdf(libraryId: libraryId, key: key)
+                } else if (try? container.decode(Bool.self, forKey: .noteKind)) == true {
+                    self = .note(libraryId: libraryId, key: key)
+                } else {
+                    throw DecodingError.dataCorrupted(DecodingError.Context(codingPath: [CodingKeys.pdfKind, CodingKeys.noteKind], debugDescription: "Item kind key not found"))
+                }
+            }
+        }
+
+        let kind: Kind
+        var userIndex: Int
+        var lastOpened: Date
+
+        init(kind: Kind, userIndex: Int, lastOpened: Date = .now) {
+            self.kind = kind
+            self.userIndex = userIndex
+            self.lastOpened = lastOpened
+        }
+    }
+
+    enum Presentation {
+        case pdf(library: Library, key: String, parentKey: String?, url: URL)
+        case note(library: Library, key: String, text: String, tags: [Tag], title: NoteEditorState.TitleData?)
+    }
+    
+    // MARK: Properties
+    private unowned let dbStorage: DbStorage
+    private unowned let fileStorage: FileStorage
+    // TODO: Use a better data structure, such as an ordered set
+    private var itemsBySessionIdentifier: [String: [Item]] = [:]
+    private var sessionIdentifierByItemKind: [Item.Kind: String] = [:]
+    private var itemsTokenBySessionIdentifier: [String: NotificationToken] = [:]
+    private var observableBySessionIdentifier: [String: PublishSubject<[Item]>] = [:]
+    private let disposeBag: DisposeBag
+
+    // MARK: Object Lifecycle
+    init(dbStorage: DbStorage, fileStorage: FileStorage) {
+        self.dbStorage = dbStorage
+        self.fileStorage = fileStorage
+        disposeBag = DisposeBag()
+    }
+    
+    // MARK: Actions
+    func observable(for sessionIdentifier: String) -> PublishSubject<[Item]> {
+        if let observable = observableBySessionIdentifier[sessionIdentifier] {
```

Comment:
Are these always accessed from main thread? Otherwise we'll need some kind of lock or queue to access them. Same with `sessionIdentifier` below.

Reply:
Yes, these are used when a view loads, a view controller is presented, or a menu action is called.